src/unistd.rs

```diff
@@ -2392,3 +2397,425 @@ pub fn access<P: ?Sized + NixPath>(path: &P, amode: AccessFlags) -> Result<()> {
+                GroupQuery::Name(name) => {
+                    unsafe {
+                        libc::getgrnam_r(CString::new(name).unwrap().as_ptr(),
+                                         &mut grp, cbuf.as_mut_ptr(), cbuf.len(), &mut res)
+                    }
+                },
+            }
+
+        };
+
+        if error == 0 {
+            if !res.is_null() {
+                Some(Ok(Group::from(res)))
+            } else {
+                None
+            }
+        } else {
+            Some(Err(Error::Sys(Errno::last())))
+        }
+    }
+}
+
+#[cfg(not(any(target_os = "android",
+              target_os = "ios",
+              target_os = "macos",
+              target_env = "musl")))]
+mod usergroupiter {
+    use libc::{self, c_char};
+    use Result;
+    use errno::Errno;
+    use super::{Error, User, Group, PWGRP_BUFSIZE};
+    use std::{mem, ptr};
+
+    /// Used to get all of the users on the system.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nix::unistd::Users;
+    /// Users::new()
+    ///     .map(|e|e.map(
+    ///         |pw| println!("{}\t{}",
+    ///                       pw.name,
+    ///                       pw.uid)))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// This iterator should not be used in different threads without synchronization; while doing so
+    /// will not cause undefined behavior, because modern systems lack re-entrant versions of
+    /// `setpwent` and `endpwent`, it is very likely that iterators running in different threads will
+    /// yield different numbers of items.
+    #[derive(Debug, Clone, Eq, PartialEq, Hash, Default)]
+    pub struct Users(usize);
+
+    impl Users {
+        /// Create a new `Users` instance with default buffer size.
+        pub fn new() -> Self {
+            unsafe { libc::setpwent(); }
+            Users(PWGRP_BUFSIZE)
+        }
+
+        /// Create a new `Users` instance with given buffer size.
+        pub fn with_bufsize(bufsize: usize) -> Self {
+            unsafe { libc::setpwent(); }
+            Users(bufsize)
+        }
+
+        /// Get the buffer size this `Users` instance was created with.
+        pub fn bufsize(&self) -> usize {
+            self.0
+        }
+    }
+
+    impl Iterator for Users {
+        type Item = Result<User>;
+        fn next(&mut self) -> Option<Result<User>> {
+
+            let mut cbuf = vec![0 as c_char; self.0];
+            let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+            let mut res = ptr::null_mut();
+
+            let error = unsafe {
+                Errno::clear();
+                libc::getpwent_r(&mut pwd, cbuf.as_mut_ptr(), self.0, &mut res)
+            };
+
+            if error == 0 && !res.is_null() {
+                Some(Ok(User::from(res)))
+            } else if error == libc::ERANGE {
+                Some(Err(Error::Sys(Errno::last())))
+            } else {
```

Comment:
I don't think you can discount the possibility of other errors.  POSIX lists `EIO`, `EMFILE`, and `ENFILE`, for example.  You should return any error you receive.